app/Http/Controllers/BeatmapTagsController.php

```diff
@@ -0,0 +1,101 @@
+<?php
+
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the GNU Affero General Public License v3.0.
+// See the LICENCE file in the repository root for full licence text.
+
+namespace App\Http\Controllers;
+
+use App\Models\Beatmap;
+use App\Models\BeatmapTag;
+use App\Models\Solo\Score;
+use App\Models\Tag;
+use Exception;
+
+class BeatmapTagsController extends Controller
+{
+    public function __construct()
+    {
+        parent::__construct();
+
+        $this->middleware('auth', [
+            'only' => [
+                'store',
+                'destroy',
+            ],
+        ]);
+
+        $this->middleware('require-scopes:public', ['only' => 'index']);
+    }
+
+    public function index(string $id)
+    {
+        $beatmapId = get_int($id);
+
+        $topBeatmapTags = \Cache::remember(
+            "beatmap_tags:{$beatmapId}",
+            $GLOBALS['cfg']['osu']['tags']['beatmap_tags_cache_interval'],
+            fn () => BeatmapTag::select(['tag_id', 'tags.name'])
+                ->selectRaw('count(*) as tag_count')
+                ->join('tags', 'beatmap_tags.tag_id', 'tags.id')
+                ->where('beatmap_id', '=', $beatmapId)
+                ->whereHas('user', function ($userQuery) {
+                    $userQuery->default();
+                })
+                ->groupBy('tag_id')
+                ->orderBy('tag_count', 'desc')
+                ->limit(50)
+                ->get()
+                ->all(),
```

Comment:
Put it in a function in one of the models; also `joinRelation` can be used like in https://github.com/ppy/osu-web/blob/019fdc3b40bac7538f401a22ae00e7446e5acbf2/app/Libraries/NotificationsBundle.php#L83
and use `qualifyColumn` instead of hardcoding the table name in the join

Reply:
what exactly is `joinRelation` going to help with here? i'm only joining `tags` to access the name of the tag, it feels like added complexity although with only that example to go by maybe i'm missing something